storage/block/result.go

```diff
@@ -88,10 +98,78 @@ type fetchBlocksMetadataResult struct {
 	blocks []FetchBlockMetadataResult
 }
 
-// newFetchBlocksMetadataResult creates new database blocks metadata
-func newFetchBlocksMetadataResult(id string, blocks []FetchBlockMetadataResult) FetchBlocksMetadataResult {
+// NewFetchBlocksMetadataResult creates new database blocks metadata
+func NewFetchBlocksMetadataResult(id string, blocks []FetchBlockMetadataResult) FetchBlocksMetadataResult {
 	return fetchBlocksMetadataResult{id: id, blocks: blocks}
 }
 
 func (m fetchBlocksMetadataResult) ID() string                         { return m.id }
 func (m fetchBlocksMetadataResult) Blocks() []FetchBlockMetadataResult { return m.blocks }
+
+type filteredBlocksMetadataIter struct {
+	start     time.Time
+	end       time.Time
+	blockSize time.Duration
+	res       []FetchBlocksMetadataResult
+
+	id       string
+	metadata Metadata
+	ri       int
+	bi       int
+}
+
+// NewFilteredBlocksMetadataIter creates a new filtered blocks metadata iterator
+func NewFilteredBlocksMetadataIter(
+	start time.Time,
+	end time.Time,
+	blockSize time.Duration,
+	res []FetchBlocksMetadataResult,
+) FilteredBlocksMetadataIter {
+	return &filteredBlocksMetadataIter{
+		start:     start,
+		end:       end,
+		blockSize: blockSize,
+		res:       res,
+	}
+}
+
+func (it *filteredBlocksMetadataIter) Next() bool {
+	if it.ri >= len(it.res) {
+		return false
+	}
+	blocks := it.res[it.ri].Blocks()
+	for it.bi < len(blocks) {
+		block := blocks[it.bi]
+		if block.Err() != nil {
+			it.bi++
+			continue
```

Comment:
Hm, wonder if this is something we want to know about, if errors are being read back for some blocks.  i.e. we might have 100% perfect checksum match rate with a peer but then there might be errors for say 50% of the blocks we try to retrieve checksums for, that would be an issue that goes unnoticed if we just skip over blocks that returned error from the peer no?